File: Week 2/problem_set_1/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

// check if a credit card is valid
int main(void)
{
    // get the credit card number from user input
    long credit_card;
    do
    {
        credit_card = get_long("Number: \n");
    }
    while (credit_card <= 0);

    // make a copy of card_num
    long card_num = credit_card;
    // sum of digits, with every other digit doubled
    int dig_sum = 0;
    // count which digit we are at. starting from the ending one
    int digit_count = 0;

    while (card_num > 0)
    {
        // debug
        //printf("card_num: %ld\n", card_num);
        //printf("digit_count: %i\n", digit_count);
        digit_count++;

        if (digit_count % 2 == 1)
        {
            // add this digit to the sum
            dig_sum += card_num % 10;
        }
        else
        {
            // add this digit * 2

            // if this digit * 2 > 9, add up the two digits of this number
            if (card_num % 10 * 2 > 9)
            {
                dig_sum += card_num % 10 * 2 - 9;
            }
            else
            {
                // else just add the digit * 2
                dig_sum += card_num % 10 * 2;
            }
        }
        //printf("digit_sum: %i\n", dig_sum);
        card_num = floor(card_num / 10);

    }

    // report INVALID
    if (dig_sum % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }

    // report the credit card company if the number is valid

    int first_digit = floor(credit_card / pow(10, digit_count - 1));
    int second_digit = floor(credit_card / pow(10, digit_count - 2));
    second_digit = second_digit % 10;

    //printf("%i\n", first_digit);
    //printf("%i\n", second_digit);
    //printf("%i\n", digit_count);
    //printf("%f\n", pow(10, digit_count - 1));
    //printf("%ld\n", credit_card);


    if (digit_count == 15 && first_digit == 3 && (second_digit == 4 || second_digit == 7))
    {
        printf("AMEX\n");
    }
    else if (digit_count == 16 && first_digit == 5 && second_digit > 0 && second_digit < 6)
    {
        printf("MASTERCARD\n");
    }
    else if ((digit_count == 13 || digit_count == 16) && first_digit == 4)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}

File: week_6/problem_set_6/credit/credit.py
from cs50 import get_int
import re
import sys

# get credit card number from user input
card = -1
while card < 0:
    card = get_int("Number: ")

# get a copy of the number as a string
card_number = str(card)

# use sum and count to keep track through the validation
digit_sum = 0
digit_count = 0

# loop through each digit from back to end
while card > 0:
    digit_count += 1
    current_digit = card % 10
    # print("digit_count:", digit_count)
    # print("current_digit: ", current_digit)

    if digit_count % 2 == 1:
        digit_sum += current_digit
    else:
        if current_digit * 2 < 9:
            digit_sum += current_digit * 2
        else:
            digit_sum += current_digit * 2 - 9

    # print("digit_sum:", digit_sum)
    card = card // 10

# invalid by digit check
if digit_sum % 10 != 0:
    print("INVALID")
    sys.exit()

## company check

# american express: 15 digits, starting with 34/37
if re.search(r'^3[47][0-9]{13}$', card_number):
    print("AMEX")
    sys.exit()
# mastercard: 16 digits, starting with 51-55
elif re.search(r'^5[1-5][0-9]{14}$', card_number):
    print("MASTERCARD")
    sys.exit()
# visa: 13/16 digits, start with 4
elif re.search(r'^4[0-9]{12}$', card_number) or re.search(r'^4[0-9]{15}$', card_number):
    print("VISA")
    sys.exit()
# invalid
else:
    print("INVALID")
    sys.exit()